src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithMetadataName.cs

```diff
@@ -137,14 +137,26 @@ public IncrementalValuesProvider<T> ForAttributeWithMetadataName<T>(
         static ImmutableArray<AttributeData> getMatchingAttributes(
             SyntaxNode attributeTarget,
             ISymbol symbol,
-            string fullyQualifiedMetadataName)
+            string fullyQualifiedMetadataName,
+            ISyntaxHelper syntaxHelper)
         {
             var targetSyntaxTree = attributeTarget.SyntaxTree;
             var result = ArrayBuilder<AttributeData>.GetInstance();
 
             addMatchingAttributes(symbol.GetAttributes());
             addMatchingAttributes((symbol as IMethodSymbol)?.GetReturnTypeAttributes());
 
+            if (symbol is INamedTypeSymbol namedTypeSymbol && syntaxHelper.IsTypeDeclaration(attributeTarget))
```

Comment:
This feels overly complicated and like it shouldn't need any new syntax helpers. We should just be able to go through the constructors, see if any have the same syntax as the attribute target, and include only that constructor's attributes.

Reply:
i think fred is saying you should check the syntax of the instanceconstructor against the syntax of the inamedtype.  if they're the same, then this is necessairly a case where you have the primary constructor.

Reply:
Ahh. I didn't think the constructor would have any syntax, since it's part of the class/record declaration rather than an independent syntactic element. But it does of course.

```csharp
if (symbol is INamedTypeSymbol namedTypeSymbol)
{
    var attrs = namedTypeSymbol
        .InstanceConstructors
        .Where(c => c.DeclaringSyntaxReferences.Any(r => r.GetSyntax() == attributeTarget))
        .SelectMany(c => c.GetAttributes())
        .ToImmutableArray();

    addMatchingAttributes(attrs);
}
```

Are we good to use `attributeTarget` here, since it's derived `symbol` in `var targetSymbol = ...` in the outer `ForAttributeWithMetadataName`? Or do we need to get the `DeclaringSyntaxReferences` from the symbol and search the cross product for a match?

Reply:
Committed and pushed but let me know if it needs further revision